deploy-tool/dns/gcloud/gcloud.go

```diff
@@ -0,0 +1,169 @@
+		Name:    fmt.Sprintf("%s.", record.Name),
+		Rrdatas: []string{record.Value},
+		Ttl:     record.TTL.Nanoseconds() / 1e9, // Expected in seconds
+		Type:    record.Type,
+		Kind:    "dns#resourceRecordSet",
+	}
+}
+
+func (p *GoogleDNSProvider) convertZone(zone string) string {
+	newZone, ok := p.zoneMappings[zone]
+	if ok {
+		return newZone
+	}
+
+	return zone
+}
+
+// AppendRecords appends DNS records to a given zone
+func (p *GoogleDNSProvider) AppendRecords(ctx context.Context, zone string, recs []libdns.Record) ([]libdns.Record, error) {
+	zone = p.convertZone(zone)
+	gcpRecords := make([]*dns.ResourceRecordSet, len(recs))
+
+	for idx, record := range recs {
+		resourceRecordSet := convertRecordType(record)
+		gcpRecords[idx] = &resourceRecordSet
+	}
+	changes := &dns.Change{
+		Additions: gcpRecords,
+	}
+	result, err := p.service.Changes.Create(p.project, zone, changes).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	if result.Status == changeStatusDone {
+		return recs, nil
+	}
+
+	changeID := result.Id
+	err = Poll("append records", 30*time.Second, 3*time.Second, func() (bool, error) {
+		result, err := p.service.Changes.Get(p.project, zone, changeID).Do()
+		if err != nil {
+			return false, err
+		}
+
+		if result.Status == changeStatusDone {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return recs, nil
+}
+
+// DeleteRecords deletes DNS records from a given zone
+func (p *GoogleDNSProvider) DeleteRecords(ctx context.Context, zone string, recs []libdns.Record) ([]libdns.Record, error) {
+	zone = p.convertZone(zone)
+	gcpRecords := make([]*dns.ResourceRecordSet, len(recs))
+
+	for idx, record := range recs {
+		resourceRecordSet := convertRecordType(record)
+		gcpRecords[idx] = &resourceRecordSet
+	}
+	changes := &dns.Change{
+		Deletions: gcpRecords,
+	}
+	result, err := p.service.Changes.Create(p.project, zone, changes).Do()
+	if err != nil {
+		return nil, err
+	}
+
+	if result.Status == changeStatusDone {
+		return recs, nil
+	}
+
+	changeID := result.Id
+	err = Poll("delete records", 30*time.Second, 3*time.Second, func() (bool, error) {
+		result, err := p.service.Changes.Get(p.project, zone, changeID).Do()
+		if err != nil {
+			return false, err
+		}
+
+		if result.Status == changeStatusDone {
+			return true, nil
+		}
+		return false, nil
+	})
+
+	if err != nil {
+		return nil, err
+	}
+
+	return recs, nil
+}
+
+// Poll is a blocking polling for a given function with a given name. The poller must return true for the polling code to stop. The poller may also return an optional error to be reported
+func Poll(job string, timeout, interval time.Duration, poller func() (bool, error)) error {
+	var lastErr error
+	timeUp := time.After(timeout)
+
+	for {
+		select {
+		case <-timeUp:
+			return fmt.Errorf("job %s timed out. Last error: %w", job, lastErr)
+		default:
+		}
+
+		stop, err := poller()
+
+		if stop {
+			return nil
+		}
+
+		if err != nil {
+			lastErr = err
```

Comment:
If the poller returns an error, should we just bail out right away?

Reply:
No the response may be an error type but the function wants it to keep retrying - if the function wants it to quit on error it could send a `true, err` as its response